### PR TITLE
feat(`claim`): genesis and types

### LINKED ISSUE #783: Initialize `claim` genesis state
Introduce the following values in the genesis state:

- `AirdropSupply` or any other name to be suggested: `claimPool`? `moduleBalance`?
- `ClaimRecords`
- `Missions` or any other name to be suggested: `Actions`? 

Implement genesis `Validate`

- `AirdropSupply` = sum of `ClaimRecord` amount
- `Missions` sum of weight = 1

### PR DESCRIPTION
**Pull request by aljo242**
Closes #783

### What does this PR does?

- add genesis of `claim` module
- create `ClaimRecord` and `Mission` types
- add `Validate` methods and tests
- minor lint and fix

### CODE REVIEW

File: proto/claim/genesis.proto [hunk 1]
```diff
@@ -2,13 +2,21 @@ syntax = "proto3";
...
 
 option go_package = "github.com/tendermint/spn/x/claim/types";
 
 // GenesisState defines the claim module's genesis state.
 message GenesisState {
-  Params params = 1 [(gogoproto.nullable) = false];
+  cosmos.base.v1beta1.Coin airdropSupply = 1 [(gogoproto.nullable) = false];
+  repeated ClaimRecord claimRecordList = 2 [(gogoproto.nullable) = false];
+  repeated Mission missionList = 3 [(gogoproto.nullable) = false];
+  uint64 missionCount = 4;
```
Comment (lubtd) [inline proto/claim/genesis.proto:18]: Why would we need count?

File: proto/claim/mission.proto [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+import "gogoproto/gogo.proto";
+
+message Mission {
+  uint64 ID = 1;
+  string description = 2;
+  string weight = 3 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
+    (gogoproto.nullable) = false];
+
```
Comment (lubtd) [inline proto/claim/mission.proto:14]:
```suggestion
```

File: proto/claim/mission.proto [hunk 3]
```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
+package tendermint.spn.claim;
+
+option go_package = "github.com/tendermint/spn/x/claim/types";
+
+import "gogoproto/gogo.proto";
+
+message Mission {
+  uint64 ID = 1;
```
Comment (lubtd) [inline proto/claim/mission.proto:9]:
```suggestion
  uint64 missionID = 1;
```

File: proto/claim/genesis.proto [hunk 4]
```diff
@@ -2,13 +2,21 @@ syntax = "proto3";
...
 // this line is used by starport scaffolding # genesis/proto/import
 
 option go_package = "github.com/tendermint/spn/x/claim/types";
 
 // GenesisState defines the claim module's genesis state.
 message GenesisState {
-  Params params = 1 [(gogoproto.nullable) = false];
+  cosmos.base.v1beta1.Coin airdropSupply = 1 [(gogoproto.nullable) = false];
+  repeated ClaimRecord claimRecordList = 2 [(gogoproto.nullable) = false];
+  repeated Mission missionList = 3 [(gogoproto.nullable) = false];
```
Comment (lubtd) [inline proto/claim/genesis.proto:17]:
```suggestion
  repeated Mission missions = 3 [(gogoproto.nullable) = false];
```

File: proto/claim/query.proto [hunk 5]
```diff
@@ -15,7 +18,31 @@ service Query {
...
+	}
+
+	// Queries a list of ClaimRecord items.
+	rpc ClaimRecordAll(QueryAllClaimRecordRequest) returns (QueryAllClaimRecordResponse) {
+		option (google.api.http).get = "/tendermint/spn/claim/claim_record";
+	}
+
+// Queries a Mission by id.
+	rpc Mission(QueryGetMissionRequest) returns (QueryGetMissionResponse) {
+		option (google.api.http).get = "/tendermint/spn/claim/mission/{ID}";
```
Comment (lubtd) [inline proto/claim/query.proto:33]:
```suggestion
		option (google.api.http).get = "/tendermint/spn/claim/mission/{missionID}";
```
If suggestion above is applied

File: proto/claim/query.proto [hunk 6]
```diff
@@ -27,4 +54,44 @@ message QueryParamsResponse {
   Params params = 1 [(gogoproto.nullable) = false];
 }
 
+message QueryGetClaimRecordRequest {
+	  string address = 1;
+
```
Comment (lubtd) [inline proto/claim/query.proto:59]:
```suggestion
```

File: proto/claim/query.proto [hunk 7]
```diff
@@ -15,7 +18,31 @@ service Query {
   rpc Params(QueryParamsRequest) returns (QueryParamsResponse) {
     option (google.api.http).get = "/tendermint/spn/claim/params";
   }
-  // this line is used by starport scaffolding # 2
+  // Queries a ClaimRecord by index.
+	rpc ClaimRecord(QueryGetClaimRecordRequest) returns (QueryGetClaimRecordResponse) {
```
Comment (lubtd) [inline proto/claim/query.proto:22]: Seems Ignite CLI doesn't format well proto files. Can we run format in the proto files?

File: x/campaign/types/special_allocations_test.go [hunk 8]
```diff
@@ -1,10 +1,12 @@
 package types_test
 
 import (
-	sdk "github.com/cosmos/cosmos-sdk/types"
 	"testing"
 
 	"github.com/stretchr/testify/require"
+
```
Comment (lubtd) [inline x/campaign/types/special_allocations_test.go:7]:
```suggestion
```

File: x/claim/client/cli/query_airdrop_supply.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/spf13/cobra"
+
+	"github.com/tendermint/spn/x/claim/types"
+)
+
+func CmdShowAirdropSupply() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "show-airdrop-supply",
+		Short: "shows AirdropSupply",
```
Comment (lubtd) [inline x/claim/client/cli/query_airdrop_supply.go:16]:
```suggestion
		Short: "shows the airdrop supply",
```

File: x/claim/client/cli/query_claim_record.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	flags.AddPaginationFlagsToCmd(cmd, cmd.Use)
+	flags.AddQueryFlagsToCmd(cmd)
+
+	return cmd
+}
+
+func CmdShowClaimRecord() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "show-claim-record [index]",
+		Short: "shows a ClaimRecord",
```
Comment (lubtd) [inline x/claim/client/cli/query_claim_record.go:49]:
```suggestion
		Short: "shows a claim record",
```

File: x/claim/client/cli/query_claim_record.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+	flags.AddPaginationFlagsToCmd(cmd, cmd.Use)
+	flags.AddQueryFlagsToCmd(cmd)
+
+	return cmd
+}
+
+func CmdShowClaimRecord() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "show-claim-record [index]",
```
Comment (lubtd) [inline x/claim/client/cli/query_claim_record.go:48]:
```suggestion
		Use:   "show-claim-record [address]",
```

File: x/claim/client/cli/query_mission.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/spf13/cobra"
+
+	"github.com/tendermint/spn/x/claim/types"
+)
+
+func CmdListMission() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list-mission",
+		Short: "list all Mission",
```
Comment (lubtd) [inline x/claim/client/cli/query_mission.go:17]:
```suggestion
		Short: "list all missions to claim airdrop",
```

File: x/claim/keeper/mission.go [hunk 13]
```diff
@@ -0,0 +1,107 @@
+package keeper
```
Comment (lubtd) [inline x/claim/keeper/mission.go:1]: I see we scaffolded mission like a list. A mission is actually identified by an ID but there will be no notion of "appending" a mission and generating the ID on the fly since the list of mission is static and defined at genesis
I suggest removing methods related to this logic:
```
GetMissionCount
SetMissionCount
AppendMission
```

**Review by lubtd — CHANGES_REQUESTED** [review]
Good progress on the module

We need to add `Validate` method for `ClaimRecord` and `Mission`

File: x/claim/types/genesis.go [hunk 14]
```diff
@@ -16,6 +20,50 @@ func DefaultGenesis() *GenesisState {
...
+		index := string(ClaimRecordKey(elem.Address))
+		claimSum = claimSum.Add(elem.Claimable)
+		if _, ok := claimRecordIndexMap[index]; ok {
+			return fmt.Errorf("duplicated index for claimRecord")
+		}
+		claimRecordIndexMap[index] = struct{}{}
+	}
+	// Check for duplicated ID in mission
+	weightSum := sdk.ZeroDec()
+	missionIDMap := make(map[uint64]bool)
```
[the remark this answers is not in the document]
Comment (aljo242) [reply x/claim/types/genesis.go:37]: This was just how it was scaffolded by `starport` lol

File: proto/claim/mission.proto (hunk 3 above)
Comment (lubtd) [reply proto/claim/mission.proto:9]: This is what we have for other types. It can help depending on the variable name

Otherwise it should be lowercase `id`

File: x/claim/types/claim_record.go [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// Validate checks the claimRecord is valid
+func (m ClaimRecord) Validate() error {
+
+	_, err := sdk.AccAddressFromBech32(m.Address)
+	if err != nil {
```
Comment (lubtd) [inline x/claim/types/claim_record.go:14]:
```suggestion
	if _, err := sdk.AccAddressFromBech32(m.Address); err != nil {
```

File: x/claim/types/mission_test.go [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+	require.NoError(t, err)
+	invalidWeightLessThanZero.Weight = amt
+
+	for _, tc := range []struct {
+		desc    string
+		mission claim.Mission
+		valid   bool
+	}{
+		{
+			desc:    "valid claim record",
```
Comment (lubtd) [inline x/claim/types/mission_test.go:30]:
```suggestion
			desc:    "valid mission",
```

Comment (lubtd) [inline x/claim/types/mission_test.go:30]: Would suggest testing edge case => `0` and `1`

File: x/claim/types/mission_test.go [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+			mission: invalidWeightGreaterThanOne,
+			valid:   false,
+		},
+		{
+			desc:    "invalid weight - less than 0",
+			mission: invalidWeightLessThanZero,
+			valid:   false,
+		},
+	} {
+		tc := tc
```
Comment (lubtd) [inline x/claim/types/mission_test.go:45]:
```suggestion
```

Why? `tc` is not a pointer

File: x/claim/types/mission_test.go [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+	"github.com/tendermint/spn/testutil/sample"
+	claim "github.com/tendermint/spn/x/claim/types"
+)
+
+func TestMission_Validate(t *testing.T) {
+	invalidWeightGreaterThanOne := sample.Mission(r)
+	amt, err := sdk.NewDecFromStr("1.0000001")
+	require.NoError(t, err)
+	invalidWeightGreaterThanOne.Weight = amt
```
Comment (lubtd) [inline x/claim/types/mission_test.go:17]: I think it would be more visual and in the idea of table test to have the full definition of mission in the case table below

For handling error from `NewDecFromStr` we could implement a `Dec` method in `testconstructor` package like we have `Shares` and `Coins`

File: x/claim/types/claim_record_test.go [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+func TestClaimRecord_Validate(t *testing.T) {
+	invalidAddress := sample.ClaimRecord(r)
+	invalidAddress.Address = "invalid"
+
+	invalidClaimable := sample.ClaimRecord(r)
+	invalidClaimable.Claimable = sdk.NewInt(-1)
+
+	invalidCompletedMissions := sample.ClaimRecord(r)
+	invalidCompletedMissions.CompletedMissions = []uint64{0, 1, 2, 0}
```
Comment (lubtd) [inline x/claim/types/claim_record_test.go:21]: Same as for `mission`

I think it's more visual to have the definition in the case table

File: x/claim/types/claim_record_test.go [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/tendermint/spn/testutil/sample"
+	claim "github.com/tendermint/spn/x/claim/types"
+)
+
+func TestClaimRecord_Validate(t *testing.T) {
+	invalidAddress := sample.ClaimRecord(r)
+	invalidAddress.Address = "invalid"
+
+	invalidClaimable := sample.ClaimRecord(r)
+	invalidClaimable.Claimable = sdk.NewInt(-1)
```
Comment (lubtd) [inline x/claim/types/claim_record_test.go:18]: For the edge case we should have a case with `0`

**Review by lubtd — COMMENTED** [review]
👍
Some comments on the test file

File: x/claim/types/mission_test.go (hunk 17 above)
Comment (aljo242) [reply x/claim/types/mission_test.go:45]: This is actually a line we have in many of our tests.  I'm not sure tbh

Comment (lubtd) [reply x/claim/types/mission_test.go:45]: https://github.com/tendermint/spn/issues/806

**Review by giunatale — CHANGES_REQUESTED** [review]
I've marked the `tc := tc` as per #806 so we can remove them from the UI. I agree with yur comments @lubtd 

rest looks good!
